src/irmin-pack/inode.ml

```diff
@@ -970,39 +1245,68 @@ struct
         let broken h =
           (* This function is called when we encounter a Broken pointer with
              Truncated layouts. *)
-          if not @@ mem h then
-            Fmt.failwith
-              "You are trying to save to the backend an inode deserialized \
-               using [Irmin.Type] that used to contain pointer(s) to inodes \
-               which are unknown to the backend. Hash: %a"
-              pp_hash h
-          else
-            (* The backend already knows this target inode, there is no need to
-               traverse further down. This happens during the unit tests. *)
-            ()
+          match index h with
+          | None ->
+              Fmt.failwith
+                "You are trying to save to the backend an inode deserialized \
+                 using [Irmin.Type] that used to contain pointer(s) to inodes \
+                 which are unknown to the backend. Hash: %a"
+                pp_hash h
+          | Some key ->
+              (* The backend already knows this target inode, there is no need to
+                 traverse further down. This happens during the unit tests. *)
+              key
         in
-        fun save_dirty arr ->
+        fun ~save_dirty arr ->
           let iter_ptr = Ptr.save ~broken ~save_dirty ~clear layout in
           Array.iter (Option.iter iter_ptr) arr
       in
       let rec aux ~depth t =
         [%log.debug "save depth:%d" depth];
         match t.v with
-        | Values _ -> add (Lazy.force t.hash) (to_bin layout t)
+        | Values _ -> (
+            let unguarded_add hash =
+              let value =
+                to_bin layout Bin.Ptr_key t
+                (* TODO: justify why this is safe *)
```

Comment:
There are not `ptr`s to convert to `Val_ref` during `to_bin`, no fail possible that way